url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/domain/singleagent/sokoban/SokobanDomain.java

```diff
@@ -0,0 +1,1345 @@
+			output = output + blocks.get(i).getDiscValForAttribute(XATTNAME) + ",";
+			output = output + blocks.get(i).getDiscValForAttribute(YATTNAME) + " ";
+		}
+
+		for (int i = 0; i < agents.size(); i++){
+			output += "agent,";
+			output = output + agents.get(i).getDiscValForAttribute(XATTNAME) + ",";
+			output = output + agents.get(i).getDiscValForAttribute(YATTNAME);
+		}
+		return output;
+	}
+
+	@Override
+	public State stringToState(String str) {
+		State st = new State();
+		int rooms = 0;
+		int doors = 0;
+		int blocks = 0;
+		int agents = 0;
+		String[] objects = str.split(" ");
+		for (int i = 0; i < objects.length; i++){
+			String[] splitobject = objects[i].split(",");
+			if (splitobject[0].equals("room")){
+				ObjectInstance room = new ObjectInstance(SOKOBANDOMAIN.getObjectClass(ROOMCLASS), ROOMCLASS + rooms);
+				room.setValue(COLORATTNAME, colors.get(Integer.parseInt(splitobject[1])));
+				room.setValue(TOPXATTNAME, Integer.parseInt(splitobject[2]));
+				room.setValue(TOPYATTNAME, Integer.parseInt(splitobject[3]));
+				room.setValue(BOTTOMXATTNAME, Integer.parseInt(splitobject[4]));
+				room.setValue(BOTTOMYATTNAME, Integer.parseInt(splitobject[5]));
+				st.addObject(room);
+				rooms++;
+			}
+			else if (splitobject[0].equals("door")){
+				ObjectInstance door = new ObjectInstance(SOKOBANDOMAIN.getObjectClass(DOORCLASS), DOORCLASS + doors);
+				door.setValue(TOPXATTNAME, Integer.parseInt(splitobject[1]));
+				door.setValue(TOPYATTNAME, Integer.parseInt(splitobject[2]));
+				door.setValue(BOTTOMXATTNAME, Integer.parseInt(splitobject[3]));
+				door.setValue(BOTTOMYATTNAME, Integer.parseInt(splitobject[4]));
+				st.addObject(door);
+				doors++;
+			}
+			else if (splitobject[0].equals("block")){
+				ObjectInstance block = new ObjectInstance(SOKOBANDOMAIN.getObjectClass(BLOCKCLASS), BLOCKCLASS + blocks);
+				block.setValue(COLORATTNAME, colors.get(Integer.parseInt(splitobject[1])));
+				block.setValue(SHAPEATTNAME, shapes.get(Integer.parseInt(splitobject[2])));
+				block.setValue(XATTNAME, Integer.parseInt(splitobject[3]));
+				block.setValue(YATTNAME, Integer.parseInt(splitobject[4]));
+				st.addObject(block);
+				blocks++;
+			}
+			else if (splitobject[0].equals("agent")){
+				ObjectInstance agent = new ObjectInstance(SOKOBANDOMAIN.getObjectClass(AGENTCLASS), AGENTCLASS + agents);
+				agent.setValue(XATTNAME, Integer.parseInt(splitobject[1].trim()));
+				agent.setValue(YATTNAME, Integer.parseInt(splitobject[2].trim()));
+				st.addObject(agent);
+				agents++;
+			}
+		}
+		createMap(st);
+		return st;
+	}
+	public static void runTrajectory(SokobanDomain constructor, Domain domain, State st, String file){
+		BufferedReader in = null;
+		String currentState = null;
+		String trajectoryString = null;
+		String command = null;
+		Trajectory trajectory = null;
+		ArrayList<State> states = new ArrayList<State>();
+
+		try {
+			in = new BufferedReader(new FileReader(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *in*. The resource is referenced at lines 1157, 1158, 1159, 1160, 1163. The resource is closed at line 1165. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *in* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).